frame/support/procedural/src/construct_runtime/expand/metadata.rs

```diff
@@ -96,6 +96,18 @@ pub fn expand_runtime_metadata(
 					#scrate::scale_info::meta_type::<#runtime>()
 				).into()
 			}
+
+			pub fn metadata_at_version(version: u32) -> Option<#scrate::OpaqueMetadata> {
+				match version {
+					// The V14 version is already implemented by `construct_runtime!` macro.
+					14 => Some(#scrate::OpaqueMetadata::new(#runtime::metadata().into())),
+					_ => None,
+				}
+			}
+
+			pub fn metadata_versions() -> #scrate::sp_std::vec::Vec<u32> {
+				#scrate::sp_std::vec![ 14 ]
+			}
```

Comment:
This should be removed.

The user should control this. We should not "dictate" what kind of metadata versions a particular runtime supports. 

I imagine that these functions exist in the `frame-metadata` pallet. Based on the enabled features of `frame-metadata` it would return the correct values. I'm open to not use some feature based approach ;)

Reply:
Yep, I like the idea of feature-based implementation, and indeed the user should decide which metadata feature they want to use :) 

> I imagine that these functions exist in the frame-metadata pallet.

Is there a `frame-metadata` pallet? 

I wonder if we could move both `metadata_at_version` and `metadata_versions` in the frame-metadata crate.

Inside the [frame-metadata](https://github.com/paritytech/frame-metadata/) crate:
-  `metadata_versions` would be easy to implement depending on feature flags
- `metadata_at_version` relies on data collected by substrate (we ll tightly couple frame-metadata with substrate):
   - [Runtime::metadata](https://github.com/paritytech/substrate/blob/f484197a8adca3625b0f102db28d3758c98cadca/frame/support/procedural/src/construct_runtime/expand/metadata.rs#L77) from substrate for V14
   - [Runtime::metadata_v15](https://github.com/paritytech/substrate/pull/13302/files#diff-2f55b49cb5adc98dbe4282ad96e4cc66adb6324abb1f122fb32e242d4963eae6R113) from substrate for V15

I think we could achieve something similar by declaring `v14` and `v15` features in `frame/support`.
When those features are present, they will activate the corresponding `frame-metadata/v14 and v15` features, this only exposes the appropriate metadata structures.
Then, we could implement `metadata_at_version` and `metadata_versions` as above in `frame/support` with feature guards.

Let me know if I understood this correctly, thanks!

Reply:
> Is there a `frame-metadata` pallet?

No, I meant the crate. Sorry.



> I think we could achieve something similar by declaring `v14` and `v15` features in `frame/support`.
> When those features are present, they will activate the corresponding `frame-metadata/v14 and v15` features, this only exposes the appropriate metadata structures.

Yeah could probably work. Depends on how it is done, but the general direction should work.

Reply:
One thing I also thought about is having some kind of "meta" metadata (haha, best naming ever :P). But the idea being that this "meta" metadata being the one generated by the FRAME macros and then we can convert this "meta" metadata into the respective versions. First I thought that the latest metadata version is probably always some superset of the versions before, but there will maybe be some point in time when this isn't true anymore.

Reply:
I like having a meta-metadata type in frame-metadata to collect everything that we need from substrate and delegate the conversion to the frame-metadata :) 

One advantage would be having a simplified code in substrate; we no longer need to collect separately the metadata fields depending on the version or handle the conversions inside substrate. This way we delegate all the work to the frame-metadata crate (work that we would have been placed here) 

I've started this PR: https://github.com/paritytech/frame-metadata/pull/49 to handle an `intermediate representation` of the metadata. 

The next steps here:
1. Get this substrate patch merged
2. Get the frame-metadata IR patch merged
3. To minimize frame-metadata releases:
 3.1.  Get frame-metadata v15 structures (+ IR conversion) patch merged 
 3.2. Release a version of frame-metadata under `v15-unstable` feature flag
4. Extend https://github.com/paritytech/substrate/pull/13302 to also construct the IR and use latest frame-metadata version  

Let me know what you think of it! Thanks!